### PR TITLE
CN0 filter redesign + review comment fixes

### PR DESCRIPTION
**Pull request by Paakkit**
This PR supersedes https://github.com/swift-nav/libswiftnav/pull/322
This PR is supposed to be merged before https://github.com/swift-nav/piksi_firmware/pull/637

### CODE REVIEW

File: src/track.c [hunk 1]
```diff
@@ -744,17 +752,22 @@ float cn0_est(cn0_est_state_t *s, float I, float Q)
...
-    float tmp = s->b * P_n / P_s;
-    s->nsr = tmp + s->xn - s->a * s->nsr;
+    float tmp = P_n / P_s;
+    float tmp2 = s->nsr;
+    s->nsr = tmp * s->b + s->xn * 2*s->b + s->xn_prev * s->b \
+                        - s->a2 * s->nsr - s->a3 * s->nsr_prev;
+    s->nsr_prev = tmp2;
+    s->xn_prev = s->xn;
     s->xn = tmp;
+
```
Comment (ljbade) [inline src/track.c]: Extra line

File: src/track.c [hunk 2]
```diff
@@ -672,14 +672,20 @@ void lock_detect_update(lock_detect_t *l, float I, float Q, float DT)
 void cn0_est_init(cn0_est_state_t *s, float bw, float cn0_0,
                   float cutoff_freq, float loop_freq)
 {
-  float Tw0 = (2*M_PI*cutoff_freq) / loop_freq;
-  s->b = Tw0 / (Tw0 + 2);
-  s->a = (Tw0 - 2) / (Tw0 + 2);
+  float Tw0 = tanf((M_PI*cutoff_freq) / loop_freq);
+  float tmp = 1.0+sqrtf(2)*Tw0+Tw0*Tw0;
+
+  s->b = Tw0*Tw0/tmp;
```
Comment (adel-mamin) [inline src/track.c]: Need to add a check for tmp to be non-zero.

File: include/libswiftnav/track.h [hunk 3]
```diff
@@ -121,11 +121,14 @@ typedef struct {
 typedef struct {
   float log_bw;     /**< Noise bandwidth in dBHz. */
   float b;          /**< IIR filter coeff. */
-  float a;          /**< IIR filter coeff. */
+  float a2;         /**< IIR filter coeff. */
```
Comment (adel-mamin) [inline include/libswiftnav/track.h]: ctypedef struct cn0_est_state_t in libswiftnav/python/swiftnav/track.pxd needs to be updated to contain these new fields. It is needed for Peregrine.

File: src/track.c [hunk 4]
```diff
@@ -672,14 +672,20 @@ void lock_detect_update(lock_detect_t *l, float I, float Q, float DT)
 void cn0_est_init(cn0_est_state_t *s, float bw, float cn0_0,
                   float cutoff_freq, float loop_freq)
 {
-  float Tw0 = (2*M_PI*cutoff_freq) / loop_freq;
-  s->b = Tw0 / (Tw0 + 2);
-  s->a = (Tw0 - 2) / (Tw0 + 2);
+  float Tw0 = tanf((M_PI*cutoff_freq) / loop_freq);
```
Comment (adel-mamin) [inline src/track.c]: loop_freq should be checked against 0.

File: src/track.c [hunk 5]
```diff
@@ -744,16 +752,25 @@ float cn0_est(cn0_est_state_t *s, float I, float Q)
...
 
     P_s = 0.5f*(I*I + s->I_prev_abs*s->I_prev_abs);
 
     s->I_prev_abs = fabsf(I);
     s->Q_prev_abs = fabsf(Q);
 
-    float tmp = s->b * P_n / P_s;
-    s->nsr = tmp + s->xn - s->a * s->nsr;
+    /* This is to avoid division by zero. */
+    if (P_s == 0.0) {
```
Comment (valeri-atamaniouk) [inline src/track.c]: P_s < EPSILON would be better

File: src/track.c (hunk 2 above)
Comment (Paakkit) [inline src/track.c]: Done

File: include/libswiftnav/track.h (hunk 3 above)
Comment (Paakkit) [inline include/libswiftnav/track.h]: Done

File: src/track.c (hunk 4 above)
Comment (Paakkit) [inline src/track.c]: Done

**Comment by Paakkit** [comment]
@gsmcmullin PR updated according to review comments

**Comment by ljbade** [comment]
As per slack, it appears that the cutoff value for too low CN0 needs to be increased to 37 dB

**Comment by valeri-atamaniouk** [comment]
Replaced by https://github.com/swift-nav/libswiftnav/pull/361

**Pull request closed without merging**